### ISSUE TITLE
Hello, just another question regarding `mll` and earlystopping.

### ISSUE THREAD

**Issue by ginward** [cornellius-gp/gpytorch#1444]
Hello, just another question regarding `mll` and earlystopping. 

If I am using approximate GP and mini-batch training, and if I am supposed to use `mll` in training mode to do earlystopping, how do I add the `mll` of different batches together? Do I just do a plain addition of all the `mll` in each mini-batch?

```
class GPModel(ApproximateGP):
    def __init__(self, inducing_points):
        variational_distribution = CholeskyVariationalDistribution(inducing_points.size(0))
        variational_strategy = VariationalStrategy(self, inducing_points, variational_distribution, learn_inducing_locations=True)
        super(GPModel, self).__init__(variational_strategy)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(lengthscale_constraint=gpytorch.constraints.GreaterThan(1)))

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
```

```
        model.train()
        likelihood.train()
        for x_batch, y_batch in minibatch_iter:
            optimizer.zero_grad()
            output = model(x_batch)
            loss = -mll(output, y_batch.t())
            minibatch_iter.set_postfix(loss=loss.item())
            loss.backward()
            optimizer.step()
            running_loss += loss.item() * batch_size
```

Or rather, do I just set the model to training model, and get the loss on the entire training dataset: 
```
        model.train()
        likelihood.train()
        train_output = model(train_x)
        train_loss = -mll(train_x, train_y).item()
```

and use `train_loss` for early stopping?

_Originally posted by @ginward in https://github.com/cornellius-gp/gpytorch/issues/1201#issuecomment-764944490_

**Comment by jacobrgardner** [comment]
If you want to do early stopping based on like a validation set nll, what you would normally do is every say `k` epochs you can process your validation set in batches and compute a standard test nll on the validation set. Then, at the end of a full training procedure you can choose the model with the best validation nll.

I doubt that doing early stopping on the _training_ mll would be very useful -- the training mll is not something you'd expect to start trending upwards. It's not exactly monotonic, but that's mostly due to fluctuations during training.

**Comment by ginward** [comment]
> If you want to do early stopping based on like a validation set nll, what you would normally do is every say `k` epochs you can process your validation set in batches and compute a standard test nll on the validation set. Then, at the end of a full training procedure you can choose the model with the best validation nll.
> 
> I doubt that doing early stopping on the _training_ loss would be very useful -- the training loss is not something you'd expect to start trending upwards. It's not exactly monotonic, but that's mostly due to fluctuations during training, and should generally be decreasing over time.

Thanks for your response. 

When I calculate the mll on the validation set, do I set the model to training model or eval mode?  From what I learned from #1201 , it seems that setting it to eval model will only return the posterior, but what we would like to maximize is the prior, right? 

Therefore, should I do the following on the validation set: 

```
        model.train()
        likelihood.train()
        train_output = model(val_x)
        train_loss = -mll(val_x, val_y).item()

```

Thanks @jacobrgardner

**Comment by ginward** [comment]
Or rather, should i simply use something like MSE, instead of NLL on the validation sample?

**Comment by basnijholt** [comment]
I am also experimenting with early stopping and my question also fits the discussion in #1445.

I wonder what the recommended method is w.r.t. regularization and preventing overfitting for simple GP regression models? I would like some early stopping condition to avoid having to tune the number of training steps.

I have tried to stop early based on the R^2 score on the validation set (instead of the loss on the validation data), which works rather well. See the plot below (x axis is training iterations).
![image](https://user-images.githubusercontent.com/6897215/107767077-b9ed5d80-6d34-11eb-96b2-971a5ea2577b.png)

However, getting training points is extremely expensive in my case, so I would prefer a solution that doesn't require extra validation data but reading your (@jacobrgardner's) message above 

>I doubt that doing early stopping on the training loss would be very useful -- the training loss is not something you'd expect to start trending upwards. It's not exactly monotonic, but that's mostly due to fluctuations during training, and should generally be decreasing over time.

makes me pessimistic about whether that is possible.

Additionally, I have tried to introduce L2 regression using the `decay_weight` parameter in `torch.optim.Adam` without much success.

Other alternatives that I can think of are to add a WhiteNoise kernel (like [`sklearn.gaussian_process.kernels.WhiteKernel`](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.WhiteKernel.html)) or by tuning the noise in the `GaussianLikelihood` likelihood function, although it is not clear to me what the effect of that should be.)

P.S. Thanks a lot for developing this package, we're seeing a speedup of ~100x in comparison with our older implementation that uses `sklearn`!

**Comment by ginward** [comment]
> I am also experimenting with early stopping and my question also fits the discussion in #1445.
> 
> I wonder what the recommended method is w.r.t. regularization and preventing overfitting for simple GP regression models? I would like some early stopping condition to avoid having to tune the number of training steps.
> 
> I have tried to stop early based on the R^2 score on the validation set (instead of the loss on the validation data), which works rather well. See the plot below (x axis is training iterations).
> ![image](https://user-images.githubusercontent.com/6897215/107767077-b9ed5d80-6d34-11eb-96b2-971a5ea2577b.png)
> 
> However, getting training points is extremely expensive in my case, so I would prefer a solution that doesn't require extra validation data but reading your (@jacobrgardner's) message above
> 
> > I doubt that doing early stopping on the training loss would be very useful -- the training loss is not something you'd expect to start trending upwards. It's not exactly monotonic, but that's mostly due to fluctuations during training, and should generally be decreasing over time.
> 
> makes me pessimistic about whether that is possible.
> 
> Additionally, I have tried to introduce L2 regression using the `decay_weight` parameter in `torch.optim.Adam` without much success.
> 
> Other alternatives that I can think of are to add a WhiteNoise kernel (like [`sklearn.gaussian_process.kernels.WhiteKernel`](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.WhiteKernel.html)) or by tuning the noise in the `GaussianLikelihood` likelihood function, although it is not clear to me what the effect of that should be.)
> 
> P.S. Thanks a lot for developing this package, we're seeing a speedup of ~100x in comparison with our older implementation that uses `sklearn`!

So far I have been using a plain-vanilla MSE on the validation set for early-stopping, which works quite well. 

As for regularisation, I found that adjusting for the noise and scale parameters, as well as the length scale parameters work quite well.

**Comment by gpleiss** [comment]
@ginward sorry for the slow response

> When I calculate the mll on the validation set, do I set the model to training model or eval mode? From what I learned from #1201 , it seems that setting it to eval model will only return the posterior, but what we would like to maximize is the prior, right?

`train()` mode. Briefly:
- `train()` mode is for when you want to compute the marginal log likelihood (e.g. the prior predictive). This is one way to measure model fit.
- `eval()` mode is for when you want to compute the posterior predictive.

> I wonder what the recommended method is w.r.t. regularization and preventing overfitting for simple GP regression models? I would like some early stopping condition to avoid having to tune the number of training steps.

> As for regularisation, I found that adjusting for the noise and scale parameters, as well as the length scale parameters work quite well.

As mentioned in #1445 - the marginal log likelihood already penalizes model complexity through the log determinant factor (see Rasmussen and Williams, Chapter 5).

Of course, it is possible that the learned noise and scale parameters aren't well fit to the data. You can add priors to these parameters (see https://docs.gpytorch.ai/en/latest/examples/00_Basic_Usage/Hyperparameters.html), which is a form of regularization.


In the future, this type of discussion would be better suited as a [Discussion topic](https://github.com/cornellius-gp/gpytorch/discussions). I'm closing this issue for now - and please open a discussion topic if you have any more questions.
